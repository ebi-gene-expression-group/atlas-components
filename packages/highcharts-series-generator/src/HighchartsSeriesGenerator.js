import seedrandom from 'seedrandom'

//https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
const getRandomArbitrary = (min, max) => Math.random() * (max - min) + min

const getRandomInt = (min, max) => {
  min = Math.ceil(min)
  max = Math.floor(max)
  return Math.floor(Math.random() * (max - min)) + min
}

const generateRandomSeriesData = (pointsCount) => {
  const seriesData = []
  for (let i = 0 ; i < pointsCount ; i++) {
    seriesData.push({
      x: getRandomArbitrary(0, 100),
      y: getRandomArbitrary(0, 100),
      expressionLevel: getRandomArbitrary(0, 10000),
      name: String(`cell_` + i)
    })
  }

  return seriesData
}

const generateRandomSeries = (seriesNames, maxPointsPerSeries, seed) => {
  seedrandom(seed, { global: true })

  return seriesNames.map((seriesName) => ({
    name: seriesName,
    data: generateRandomSeriesData(getRandomInt(0, maxPointsPerSeries))
  }))
}

const HighchartsSeriesGenerator = {
  generate: generateRandomSeries
}

export default HighchartsSeriesGenerator
