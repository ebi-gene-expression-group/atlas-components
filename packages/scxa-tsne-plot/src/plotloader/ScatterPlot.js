import React from 'react'
import PropTypes from 'prop-types'
import ReactHighcharts from 'react-highcharts'

import HighchartsExporting from 'highcharts/modules/exporting'
import HighchartsHeatmap from 'highcharts/modules/heatmap'
import highchartsAdaptChartToLegendModule from 'highcharts-adapt-chart-to-legend'

//import HighchartsMap from 'highcharts/modules/map'
//import highchartsYAxisPanningModule from './modules/y-axis-panning'

import HighchartsBoost from './modules/boost'
import HighchartsExportStyle from './modules/export-style'
import highchartsHeatmapLegendModule from './modules/heatmap-legend'

import deepmerge from 'deepmerge'
import SeriesPropTypes from './SeriesPropTypes'

const Highcharts = ReactHighcharts.Highcharts
// Only include modules if Highcharts isn’t a *good* mock -- Boost/Exporting can break tests
// if (Highcharts.getOptions()) {...}
HighchartsExporting(Highcharts)
HighchartsBoost(Highcharts)
HighchartsHeatmap(Highcharts)
highchartsHeatmapLegendModule(Highcharts, `expressionLevel`)
highchartsAdaptChartToLegendModule(Highcharts)
HighchartsExportStyle(Highcharts)

// To add drag-to-pan functionality:
// HighchartsMap(Highcharts)
// highchartsYAxisPanningModule(Highcharts)

const getRadiusSize = (totalNumberOfPoints) => {
  if (totalNumberOfPoints >= 40000) {
    return 1
  } else if (totalNumberOfPoints >= 10000) {
    return 2
  } else if (totalNumberOfPoints >= 5000) {
    return 3
  }
  return 4
}

const highchartsBaseConfig = {
  credits: {
    enabled: false
  },

  chart: {
    type: `scatter`,
    spacingLeft: 0,
    height: `100%`,
    panning: true,
    spacingTop: 50,
    zoomType: `xy`,
    plotBackgroundColor: `transparent`, // This needs to be set to allow access to this.plotBackground
    events: {
      load: function() { this.plotBackground.htmlCss({ cursor: `crosshair` }) }
    }
  },

  legend: {
    adjustChartSize: true,
    navigation: {
      enabled: false
    }
  },

  // mapNavigation: {
  //   enabled: true,
  //   enableMouseWheelZoom: false,
  //   buttonOptions: {
  //     theme: {
  //       fill: `white`,
  //       'stroke-width': 1,
  //       stroke: `silver`,
  //       r: 0,
  //       states: {
  //           hover: {
  //               fill: `#a4edba`
  //           },
  //           select: {
  //               stroke: `#039`,
  //               fill: `#a4edba`
  //           }
  //       }
  //     },
  //     verticalAlign: `bottom`
  //   }
  // },

  title: {
    text: null,
    style: {
      fontSize: `25px`,
      fontWeight: `bold`
    }
  },
  xAxis: {
    lineColor: `#e6e6e6`,
    title: {
      text: null
    },
    labels: {
      enabled: false
    },
    tickWidth: 0
  },
  yAxis: {
    lineColor: `#e6e6e6`,
    title: {
      text: null
    },
    labels: {
      enabled: false
    },
    gridLineWidth: 0,
    lineWidth: 1,
    endOnTick: false,
    startOnTick: false
  },
  colors: [`#b25fbc`, `#76b341`, `#6882cf`, `#ce9b44`, `#c8577b`, `#4fae84`, `#c95c3f`, `#7c7f39`],
  plotOptions: {
    series: {
      turboThreshold: 0,
      animation: false
    }
  },

  navigation: {
    buttonOptions: {
      theme: {
        style: {
          fontSize: `15px`
        },
        states: {
          select: {
            style: {
              fontWeight: `normal`,
              color: `black`
            }
          }
        }
      }
    },
    menuItemStyle: {
      fontSize: `15px`
    }
  },

  exporting: {
    buttons: {
      contextButton: {
        text: `<i class="icon icon-functional" data-icon="="></i>&nbsp;Download`,
        symbol: null
      }
    }
  }
}

const ScatterPlot = (props) => {
  const { chartClassName, series, highchartsConfig, legendWidth } = props
  const boostThreshold = 10000
  const totalNumberOfPoints = series.reduce((acc, cur) => acc + cur.data.length, 0)

  const config =
    deepmerge.all([
      highchartsBaseConfig,
      {
        boost: {
          useGPUTranslations: true,
          usePreAllocated: true,
          seriesThreshold: totalNumberOfPoints >= boostThreshold ? 1 : null
        }
      },
      {
        plotOptions: {
          series: {
            boostThreshold: boostThreshold,
            marker: {
              radius: getRadiusSize(totalNumberOfPoints)
            },
            stickyTracking: false
          }
        }
      },
      {
        series: series
      },
      highchartsConfig,
      {
        legend: {
          symbolWidth: legendWidth
        }
      }
    ], {
      arrayMerge: (destination, source) => source // Don’t merge, overwrite instead
    })

  return (
    <div className={chartClassName}>
      <ReactHighcharts config={config}/>
    </div>
  )
}

ScatterPlot.propTypes = {
  chartClassName: PropTypes.string,
  series: SeriesPropTypes,
  highchartsConfig: PropTypes.object,
  legendWidth: PropTypes.number
}

ScatterPlot.defaultProps = {
  highchartsConfig: {}
}

export default ScatterPlot
