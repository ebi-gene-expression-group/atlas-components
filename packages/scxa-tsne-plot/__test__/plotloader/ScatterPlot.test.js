import React from 'react'
// Highcharts can only be shallow-rendered unless it’s mocked, see __mocks__/highcharts.js
import Enzyme from 'enzyme'
import {shallow, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import HighchartsSeriesGenerator from 'highcharts-series-generator'

import ScatterPlot from '../../src/plotloader/ScatterPlot'

Enzyme.configure({ adapter: new Adapter() })

describe(`ScatterPlot`, () => {

  test(`merges additional options for Highcharts`, () => {
    const highchartsConfig = {
      chart: {
        height: `50%`, // existing property, overwritten
        width: `50%`   // new property
      }
    }
    const wrapper = mount(<ScatterPlot series={[]} highchartsConfig={highchartsConfig}/>)
    const highchartsWrapper = wrapper.find(`HighchartsChart`)
    expect(highchartsWrapper.prop(`config`)).toHaveProperty(`chart.height`, `50%`)
    expect(highchartsWrapper.prop(`config`)).toHaveProperty(`chart.width`, `50%`)
  })

  test(`with no series matches snapshot`, () => {
    const tree = shallow(<ScatterPlot series={[]}/>)
    expect(tree).toMatchSnapshot()
  })

  test(`marker radius changes depending on number of total points`, () => {
    const wrapper = mount(<ScatterPlot series={[]}/>)

    const longSeriesName = `Series with 5,000 points`
    const longSeriesData = []
    for (let i = 0 ; i < 5000 ; i++) {
      longSeriesData.push({
        name: `Point ${i}`,
        x: 0,
        y: 0,
        expressionLevel: 0
      })
    }
    const longSeries = {
      name: longSeriesName,
      data: longSeriesData
    }
    wrapper.setProps({ series: [longSeries] }).mount()
    const markerRadiusLongSeries = wrapper.find(`HighchartsChart`).prop(`config`).plotOptions.series.marker.radius

    const shortSeriesName = `Series with 4,999 points`
    const shortSeriesData = []
    for (let i = 0 ; i < 4999 ; i++) {
      shortSeriesData.push({
        name: `Point ${i}`,
        x: 0,
        y: 0,
        expressionLevel: 0
      })
    }
    const shortSeries = {
      name: shortSeriesName,
      data: shortSeriesData
    }
    wrapper.setProps({ series: [shortSeries] }).mount()
    const markerRadiusShortSeries = wrapper.find(`HighchartsChart`).prop(`config`).plotOptions.series.marker.radius

    expect(markerRadiusLongSeries).toBeLessThan(markerRadiusShortSeries)
  })

  test(`matches snapshot with randomized series`, () => {
    const seed = `A hair, Morty. I need one of your hairs. This isn’t Game of Thrones.`
    const seriesNames = [`Series 1`, `Series 2`, `Series 3`, `Series 4`, `Series 5`]
    const maxPointsPerSeries = 1000
    const series = HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries, seed)
    const tree = shallow(<ScatterPlot series={series}/>)
    expect(tree).toMatchSnapshot()
  })

  test(`doesn’t check for unique names in the same series or among different series`, () => {
    const series = {
      series: [
        {
          name: `Cluster 1`,
          data: [
            {x: 0, y: 0, name: `Cell A`, expressionLevel: 0},
            {x: 0, y: 0, name: `Cell A`, expressionLevel: 0},
            {x: 1.11111, y: 2.22222, name: `Cell B`, expressionLevel: 9.99999}
          ]
        },
        {
          name: `Cluster 2`,
          data: [
            {x: 0, y: 0, name: `Cell A`, expressionLevel: 0},
            {x: 1.11111, y: 2.22222, name: `Cell B`, expressionLevel: 9.99999}
          ]
        },
        {
          name: `Cluster 3`,
          data: [
            {x: 1.11111, y: 2.22222, name: `Cell A`, expressionLevel: 9.99999}
          ]
        }
      ]
    }

    const tree = shallow(<ScatterPlot {...series}/>)
    expect(tree).toMatchSnapshot()
  })
})
