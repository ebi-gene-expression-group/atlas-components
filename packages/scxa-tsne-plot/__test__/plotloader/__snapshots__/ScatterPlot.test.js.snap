// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScatterPlot doesnâ€™t check for unique names in the same series or among different series 1`] = `ShallowWrapper {}`;

exports[`ScatterPlot matches snapshot with randomized series 1`] = `ShallowWrapper {}`;

exports[`ScatterPlot with no series matches snapshot 1`] = `ShallowWrapper {}`;
